shortcut in reading nested group members when IN_CHAIN is available

blizzz opened the pull request:
Currently, to support members in nested groups, we cycle through them. Some LDAP server do implement the LDAP_MATCHING_RULE_IN_CHAIN (aka 1.2.840.113556.1.4.1941) extensible matcher (AD, Samba 4), which would speed up reading of group members drastically.

This matcher can not be used blindly, as server that do not implement it, return empty results. To avoid testing every time, the state is saved in the configuration, so that when working in servers that do not support this mechanism, it is not being attempted all over again.

blizzz committed:
LDAP: shortcut in reading nested group members when IN_CHAIN is available

Signed-off-by: Arthur Schiwon <blizzz@arthur-schiwon.de>

blizzz commented:
/backport to stable19

blizzz commented:
/backport to stable18

ChristophWurst reviewed (approved):
:elephant:

juliushaertl commented on apps/user_ldap/lib/Configuration.php:
```diff
@@ -110,6 +110,7 @@ class Configuration {
 		'ldapDynamicGroupMemberURL' => null,
 		'ldapDefaultPPolicyDN' => null,
 		'ldapExtStorageHomeAttribute' => null,
+		'ldapMatchingRuleInChainState' => 'unknown',
```
I'd love to see the magic strings unknown/unavailable/available moved into constants, otherwise :+1:

MorrisJobke commented:
Let's wait for #16737 first and then rebase this one here.

blizzz commented:
> Let's wait for #16737 first and then rebase this one here.

@MorrisJobke ping :)

blizzz commented:
ah wait, while looking at another issue where this can help with, i found some more improvments

blizzz committed:
do not flip available state to unavailable, allow empty results

- the detection relies that the first, requested result is not empty
- it might be empty though – groups without members
- protect switching from available to unavailable
  - switching the other way around was also not envisaged either

Signed-off-by: Arthur Schiwon <blizzz@arthur-schiwon.de>

MorrisJobke reviewed (approved):
Code looks good 👍
